src/main/java/com/github/lunatrius/schematica/world/schematic/SchematicStructure.java

```diff
@@ -0,0 +1,113 @@
+import com.github.lunatrius.schematica.api.ISchematic;
+import com.github.lunatrius.schematica.nbt.NBTHelper;
+import com.github.lunatrius.schematica.reference.Names;
+import com.github.lunatrius.schematica.reference.Reference;
+import com.github.lunatrius.schematica.world.storage.Schematic;
+
+public class SchematicStructure extends SchematicFormat {
+    @Override
+    public ISchematic readFromNBT(final NBTTagCompound tagCompound) {
+        final ItemStack icon = SchematicUtil.getIconFromNBT(tagCompound);
+
+        final Template template = new Template();
+        template.read(tagCompound);
+
+        final Schematic schematic = new Schematic(icon,
+                template.size.getX(), template.size.getY(), template.size.getZ(), template.getAuthor());
+
+        for (Template.BlockInfo block : template.blocks) {
+            schematic.setBlockState(block.pos, block.blockState);
+            if (block.tileentityData != null) {
+                try {
+                    // This position isn't included by default
+                    block.tileentityData.setInteger("x", block.pos.getX());
+                    block.tileentityData.setInteger("y", block.pos.getY());
+                    block.tileentityData.setInteger("z", block.pos.getZ());
+
+                    final TileEntity tileEntity = NBTHelper.readTileEntityFromCompound(block.tileentityData);
+                    if (tileEntity != null) {
+                        schematic.setTileEntity(block.pos, tileEntity);
+                    }
+                } catch (final Exception e) {
+                    Reference.logger.error("TileEntity failed to load properly!", e);
+                }
+            }
+        }
+
+        // for (Template.EntityInfo entity : template.entities) {
+        //     schematic.addEntity(...);
+        // }
+
+        return schematic;
+    }
+
+    @Override
+    public boolean writeToNBT(final NBTTagCompound tagCompound, final ISchematic schematic) {
+        Template template = new Template();
+        template.size = new BlockPos(schematic.getWidth(), schematic.getHeight(), schematic.getLength());
+
+        template.setAuthor(schematic.getAuthor());
+
+        // NOTE: Can't use MutableBlockPos here because we're keeping a reference to it in BlockInfo
+        for (BlockPos pos : BlockPos.getAllInBox(BlockPos.ORIGIN, template.size.add(-1, -1, -1))) {
+            final TileEntity tileEntity = schematic.getTileEntity(pos);
+            final NBTTagCompound compound;
+            if (tileEntity != null) {
+                compound = NBTHelper.writeTileEntityToCompound(tileEntity);
+                // Tile entities in structures don't store these coords
+                compound.removeTag("x");
+                compound.removeTag("y");
+                compound.removeTag("z");
+            } else {
+                compound = null;
+            }
+
+            template.blocks.add(new Template.BlockInfo(pos, schematic.getBlockState(pos), compound));
```

Comment:
Just a random though.

Filtering out air blocks could significantly cut down on the file size of the schematics that are mostly air blocks. Only question is, do/should we care for any blocks that pretend to be air blocks (like the Railcraft heat blocks)?